types/stylelint/index.d.ts

```diff
@@ -63,25 +83,74 @@ export type RuleOption = {
     optional: true;
 };
 
+export type RuleMessageValue = string | ((...args: any[]) => string);
+
 export namespace utils {
     function report(violation: {
         ruleName: string;
-        result: LintResult;
+        result: postcss.Result;
         message: string;
-        node: any;
+        node: postcss.Node;
         index?: number;
         word?: string;
         line?: number;
     }): void;
 
-    function ruleMessages(ruleName: string, messages: { [key: string]: any; }): typeof messages;
+    function ruleMessages<T extends {[key: string]: RuleMessageValue}>(
+        ruleName: string,
+        messages: T): T;
 
-    function validateOptions(result: LintResult, ruleName: string, ...options: RuleOption[]): boolean;
+    function validateOptions(result: postcss.Result, ruleName: string,
+        ...options: RuleOption[]): boolean;
 
-    function checkAgainstRule(options: { ruleName: string; ruleSettings: any; root: any; }, callback: (warning: string) => void): void;
+    function checkAgainstRule(options: {
+        ruleName: string;
+        ruleSettings: any;
+        root: any;
+    }, callback: (warning: string) => void): void;
 }
 
+export type Plugin = (primaryOption: any, secondaryOptions: RuleOption[]) =>
```

Comment:
this is and always was wrong, will fix tomorrow if i can (its not a `RuleOption`, what we call a `RuleOption` is in fact a validateOptions assertion object).